applications/launchpad/backend/src/grpc/base_node_grpc_client.rs

```diff
@@ -0,0 +1,131 @@
+    wallet_client::WalletClient,
+    Empty,
+    GetBalanceRequest,
+    GetBalanceResponse,
+    GetIdentityRequest,
+    GetIdentityResponse,
+    TransactionEvent,
+    TransactionEventRequest,
+    TransactionEventResponse,
+};
+use tauri::{async_runtime::block_on, http::status};
+use tokio::task;
+use tonic::transport::Channel;
+
+use super::{error::GrpcError, BlockStateInfo};
+use crate::{
+    docker::{DockerWrapperError, LaunchpadConfig, BASE_NODE_GRPC_ADDRESS_URL},
+    error::LauncherError,
+};
+
+type Inner = BaseNodeClient<tonic::transport::Channel>;
+
+#[derive(Clone)]
+pub struct GrpcBaseNodeClient {
+    inner: Option<Inner>,
+}
+
+impl GrpcBaseNodeClient {
+    pub fn new() -> GrpcBaseNodeClient {
+        Self { inner: None }
+    }
+
+    pub async fn connection(&mut self) -> Result<&mut Inner, GrpcError> {
+        if self.inner.is_none() {
+            let inner = Inner::connect(BASE_NODE_GRPC_ADDRESS_URL).await?;
+            self.inner = Some(inner);
+        }
+        self.inner
+            .as_mut()
+            .ok_or_else(|| GrpcError::FatalError("no connection".into()))
+    }
+
+    pub async fn connected(&mut self) -> bool {
+        loop {
+            match self.connection().await {
+                Ok(_) => {
+                    info!("#### Connected....");
+                    break;
+                },
+                Err(_) => {
+                    sleep(Duration::from_secs(3));
+                    info!("---> Waiting for base node....");
+                    continue;
+                },
+            }
+        }
+
+        true
+    }
+
+    pub async fn stream(&mut self) -> Result<impl Stream<Item = BlockStateInfo>, GrpcError> {
+        let (mut sender, receiver) = mpsc::channel(100);
+        let connection = self.connection().await.unwrap().clone();
+        task::spawn(async move {
+            loop {
+                let request = Empty {};
+                let response = match connection.clone().get_sync_progress(request).await {
+                    Ok(response) => response.into_inner(),
+                    Err(status) => {
+                        error!("Failed reading progress from base node: {}", status);
+                        return;
+                    },
+                };
+
+                info!("Response: {:?}", response);
+
+                match response.clone().state() {
+                    tari_app_grpc::tari_rpc::SyncState::Done => {
+                        info!("GONGRATS....Base node is synced.");
+                        return;
+                    },
+                    tari_app_grpc::tari_rpc::SyncState::Header | tari_app_grpc::tari_rpc::SyncState::Block => {
+                        sender.try_send(BlockStateInfo::from(response)).unwrap()
+                    },
+                    sync_state => info!("Syncing is being started. Current state: {:?}", sync_state),
+                }
+                sleep(Duration::from_secs(10));
```

Comment:
Don't use `thread::sleep`

Reply:
fixed :) I haven't known it is a bad practice.